contracts/root/StakeManager.sol

```diff
@@ -37,118 +30,134 @@ contract StakeManager is StakeManagerInterface, RootChainable, Lockable {
 
+  uint96 MAX_UINT96 = (2**96)-1; //Todo: replace with erc20 token max value
+
+  ERC20 public token;
+
   // genesis/governance variables
-  uint256 public dynasty = 2**13;  // unit: epoch
+  uint256 public DYNASTY = 2**13;  // unit: epoch
   uint256 public MIN_DEPOSIT_SIZE = (10**18);  // in ERC20 token
   uint256 public EPOCH_LENGTH = 256; // unit : block
-  uint256 public WITHDRAWAL_DELAY = dynasty.div(2); // unit: epoch
+  uint256 public WITHDRAWAL_DELAY = DYNASTY.div(2); // unit: epoch
+  uint256 public UNSTAKE_DELAY = DYNASTY.mul(2); // unit: epoch
 
-  uint256 public validatorThreshold = 10;
+
+  uint256 public validatorThreshold = 10; //128
   uint256 public maxStakeDrop = 95; // in percent 100-x, current is 5%
-  uint256 public minLockInPeriod = 2; // unit: dynasty
+  uint256 public minLockInPeriod = 2; // unit: DYNASTY
   uint256 public totalStaked = 0;
   uint256 public currentEpoch = 1;
 
-  AvlTree validatorList;
-
-  struct Staker {
+  struct Validator {
     uint256 epoch;
     uint256 amount;
+    uint256 reward;
     uint256 activationEpoch;
     uint256 deactivationEpoch;
     address signer;
+    uint96 validatorId;
   }
 
-  // signer to Staker mapping
-  mapping (address => address) public signerToStaker;
-
-  mapping (address => Staker) public stakers;
-
   struct State {
     int256 amount;
     int256 stakerCount;
   }
+
+  // signer to Validator mapping
+  mapping (address => address) public signerToValidator;
+  mapping (uint96 => address) private idToValidator; // for internal use only, for accurate data use validators mapping
+
+  mapping (address => Validator) public validators;
+
+
   //Mapping for epoch to totalStake for that epoch
   mapping (uint256 => State) public validatorState;
 
   constructor () public {
-    validatorList = new AvlTree(); // TODO: bind with stakemanager
   }
 
   // only staker
-  modifier onlyStaker() {
-    require(totalStakedFor(msg.sender) > 0);
+  modifier onlyStaker(uint256 validatorId) {
+    require(ownerOf(validatorId) == msg.sender);
     _;
   }
 
-  function stake(address unstakeValidator, address signer, uint256 amount) public {
-    stakeFor(msg.sender, unstakeValidator, signer, amount);
+  function stake(uint256 amount, address signer, uint96 validatorId) public {
+    stakeFor(msg.sender, amount, signer, validatorId);
   }
 
-  function stakeFor(address user, address unstakeValidator, address signer, uint256 amount) public onlyWhenUnlocked {
-    require(stakers[user].epoch == 0, "No second time staking");
-    // currentValidatorSetSize*2 means everyone is commited
-    require(validatorThreshold*2 > validatorList.currentSize(), "Validator set full");
-    require(amount < MAX_UINT96, "Stay realistic!!");
+  function stakeFor(address user, uint256 amount, address signer, uint96 validatorId) public onlyWhenUnlocked {
+    require(validators[user].epoch == 0, "No second time staking");
+    require(validatorId > 0 && validatorId <= validatorThreshold);
```

Comment:
I was hoping validator Id will be different than total validators. Sometimes we validator exits, it vacates the seat but can still have validatorId till it gets rewards (and during withdraw delay period)

Reply:
now every validator has unique ValidatorId so no conflicts while syncing etc.